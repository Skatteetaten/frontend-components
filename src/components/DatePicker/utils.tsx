export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
