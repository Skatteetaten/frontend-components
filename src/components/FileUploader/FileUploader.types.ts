import { calloutState, LabelWithCalloutProps } from '../LabelWithCallout';

export enum Language {
  en = 'en',
  nb = 'nb',
  nn = 'nn',
}

export enum FileFormatTypes {
  doc = '.doc',
  docx = '.docx',
  jpeg = '.jpeg',
  jpg = '.jpg',
  pdf = '.pdf',
  png = '.png',
  tif = '.tif',
  txt = '.txt',
  xml = '.xml',
  csv = '.csv',
  odf = '.odf',
  odp = '.odp',
  ods = '.ods',
  odt = '.odt',
  ppt = '.ppt',
  pptx = '.pptx',
  xls = '.xls',
  xlsx = '.xlsx',
}

export interface AttachmentMetadata extends File {
  id: string;
}
/*
 * visibleName FileUploader (Filopplaster)
 */
export interface FileUploaderProps {
  /** Akksepterte filformater */
  acceptedFileFormats?: Array<FileFormatTypes>;
  /** Tekst for aksepeterte typer*/
  acceptedFileFormatsLabel?: string;
  /** Tekst for opplastingskomponenten */
  addFileString?: string | JSX.Element;
  /** Funksjon som kjøres etter opplasting */
  afterUpload?: (uploadedFiles: any) => void;
  /** aria-label @deprecated */
  ariaLabel?: string;
  /** string for Apikall */
  axiosPath?: string;
  /** CSS class */
  className?: string;
  /** trigger funksjon til å slette alle filer */
  deleteAllFiles?: boolean;
  /** Aria-label for "fjern fil"-knapp */
  deleteButtonAriaLabel?: string;
  /** Funksjon for å slette opplastet fil */
  deleteFile?: (file: AttachmentMetadata | File, errors: string[]) => void;
  /**feilmelding for oversteget av filstørrelsesgrense*/
  exceedFileSizeLimitErrorMessage?: string;
  /** Opplastede filer */
  files?: Array<any>;
  /** Størrelsesgrense til en enkelt fil i bit*/
  fileSizeLimit?: number;
  /** Forsinkelse før opplasting i millisekunder*/
  forsinkelse?: number;
  /** Hjelpetekst */
  help?: string | JSX.Element;
  /** Id - should have a value so unique references to labels inside component can be made */
  id?: string;
  /** Tilleggsinformasjon */
  info?: string | JSX.Element;
  /** Definer ugyldige tegn som skal erstattes med "_". NormalizeFileName vil erstatte alle ugyldige tegn dersom denne verdien ikke er satt. */
  invalidCharacterRegexp?: RegExp;
  /**Funksjon som kjører dersom spinner forandrer state */
  isLoading?: (loading: boolean) => void;
  /** Beskrivende label for FileUploader */
  label?: string;
  /** aria-label for knapp i label */
  labelButtonAriaLabel?: string;
  /** Overstyr label, se LabelWithCallout komponent */
  labelWithCalloutProps?: LabelWithCalloutProps;
  /** Language selection for what the screen reader reads out. Default is Norwegian Bokmål */
  language?: Language;
  /** Spinner når fil laster */
  loading?: boolean;
  /** Mulighet for å laste opp flere filer */
  multipleFiles?: boolean;
  /** Erstatter tegn som er ugyldige */
  normalizeFileName?: boolean;
  /** Brukerspesifisert event for callout */
  onCalloutToggle?: (
    oldCalloutState: calloutState,
    newCalloutState: calloutState
  ) => void;
  /** Parametere for axioskall */
  queryParams?: any;
  /** Om merke for obligatorisk felt skal vises */
  required?: boolean;
  /** Funksjon for filopplasting */
  uploadFile?: (file: File) => void;
  /** Gjør at DELETE operasjonen, ved slett av opplastet fil, fungerer når løsningen kjører bak WebSeal.
   * Default implementasjon legger ved en tom body i DELETE requesten som er nødvendig for løsninger som kjører bak BigIp
   *  **/
  usesWebSealCompatibleDelete?: boolean;
  /**
   * Funksjon for filnedlasting
   * */
  downloadFile?: (file: File) => void;
}
