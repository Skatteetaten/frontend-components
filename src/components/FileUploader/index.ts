export { default } from './FileUploader';
export * from './FileUploader';
