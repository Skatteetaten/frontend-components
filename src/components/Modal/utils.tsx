import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
