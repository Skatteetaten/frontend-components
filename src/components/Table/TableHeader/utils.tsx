import { TFunction } from 'react-i18next';

export const getIconNameForTheadCell = (
  isSorted: boolean,
  isSortedAscending: boolean
) => {
  if (isSorted) {
    return isSortedAscending ? 'ArrowDown' : 'ArrowUp';
  }

  return 'ArrowUpDown';
};

export const getAriaLabelForTheadCell = (
  name: string | JSX.Element,
  fieldName: string,
  isSorted: boolean,
  isSortedAscending: boolean,
  t: TFunction<Array<string>>
) => {
  let ariaLabel = fieldName;
  if (typeof name === 'string') {
    ariaLabel = isSorted
      ? isSortedAscending
        ? name.concat(' ', t('table.sorted_ascending'))
        : name.concat(' ', t('table.sorted_descending'))
      : name.concat(' ', t('table.sortable'));
  }

  return ariaLabel;
};
