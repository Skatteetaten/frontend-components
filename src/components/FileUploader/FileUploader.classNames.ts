import { mergeStyleSets } from '@fluentui/merge-styles';
import { getTheme } from '@fluentui/react';
import { FontSizes, LineHeightSizes, PaletteProps } from '../utils';
import { FileUploaderProps } from './FileUploader.types';

export const getClassNames = (props: FileUploaderProps) => {
  const palette = getTheme().palette as PaletteProps;

  return mergeStyleSets({
    srOnly: {
      position: 'absolute',
      width: '1px',
      height: '1px',
      padding: '0',
      margin: '-1px',
      overflow: 'hidden',
      clip: 'rect(0, 0, 0, 0)',
      whiteSpace: 'nowrap',
      borderWidth: '0',
    },
    acceptedFileFormats: {
      fontWeight: 'bold',
    },
    informationWrapper: {
      marginTop: '8px',
      fontSize: FontSizes.xSmall,
      lineHeight: LineHeightSizes.xSmall,
      color: palette.skeColor.darkGrey,
    },
    errorColor: {
      color: `${palette.skeColor.error} !important`,
    },
    fileList: {
      listStyle: 'none !important',
      padding: 0,
      selectors: {
        i: {
          color: palette.skeColor.blue,
          float: 'right',
        },
        li: {
          wordBreak: 'break-all',
          display: 'block',
        },
      },
    },
    fileListWrapper: {
      display: 'grid',
    },
    fileListCancelBtn: {
      backgroundColor: 'transparent',
      float: 'right',
      padding: 0,
      border: 0,
      borderStyle: 'none',
      cursor: 'pointer',
      marginTop: 4,
      fontSize: '20px',
    },
    fileIcon: {
      fontSize: '20px',
    },
    fileName: {
      float: 'left',
      width: '90%',
      display: 'inline',
      lineHeight: '1.4',
      selectors: {
        i: {
          float: 'left',
          marginRight: '8px',
          color: palette.skeColor.grey,
        },
        span: {
          display: 'flex',
          marginTop: '3px',
        },
      },
    },
    fileUploadInput: {
      opacity: 0,
      overflow: 'hidden',
      position: 'absolute',
      zIndex: -1,
    },
    main: {},
    uploadArea: {
      border: `2px dashed ${palette.skeColor.blue}`,
      backgroundColor: palette.skeColor.lightBlue,
      padding: '16px',
      textAlign: 'center',
      marginTop: '8px',
      marginBottom: '8px',
      cursor: 'pointer',
      transition: 'all 80ms ease-in-out 0s',
      selectors: {
        ':hover': {
          border: `2px dashed ${palette.skeColor.blue}`,
          transform: 'scale(1.02) translateY(-1px)',
          boxShadow: 'rgba(0, 0, 0, 0.3) 0px 10px 6px -10px',
        },
        ':focus': {
          border: `2px solid ${palette.skeColor.blue}`,
          transform: 'scale(1.02) translateY(-1px)',
          boxShadow: 'rgba(0, 0, 0, 0.3) 0px 10px 6px -10px',
        },
        ':active': {
          border: `2px solid ${palette.skeColor.blue}`,
          transform: 'scale(1) translateY(0)',
          boxShadow: 'rgba(0, 0, 0, 0.0) 0px 0 0 0',
        },
      },
    },
    uploadAreaIcon: {
      fontSize: FontSizes.icon,
      color: palette.skeColor.blue,
    },
  });
};
