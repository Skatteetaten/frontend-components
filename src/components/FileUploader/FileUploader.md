** FileUploader **

Komponenten er basert på at du sender inn sti til API. Dersom ikke dette skulle passe er det mulighet for å bruke uploadFile og kjøre en egendefinert funksjon.

```js
import React from 'react';
import FileUploader, {
  FileFormatTypes
} from '@skatteetaten/frontend-components/FileUploader';
const [files, setFiles] = React.useState([]);
const [spinner, setSpinner] = React.useState(false);

<div style={{ width: '300px' }}>
  <FileUploader
    label={'Last opp vedlegg'}
    help="Tekst som hjelper brukeren å fylle ut feltet."
    acceptedFileFormats={[
      FileFormatTypes.doc,
      FileFormatTypes.docx,
      FileFormatTypes.txt
    ]}
    files={files}
    uploadFile={file => {
      setSpinner(true);
      setTimeout(() => {
        const newList = [...files];
        newList.push(file);
        setFiles(newList);
        setSpinner(false);
      }, 2000);
    }}
    deleteFile={file => {
      const newList = files.filter(fileInList => fileInList !== file);
      setFiles(newList);
    }}
    loading={spinner}
  />
</div>;
```
