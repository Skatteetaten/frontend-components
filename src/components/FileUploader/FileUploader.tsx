import * as React from 'react';
import axios, { AxiosResponse } from 'axios';
import { getClassNames } from './FileUploader.classNames';
import classnames from 'classnames';
import Icon from '../Icon';
import LabelWithCallout, {
  calloutState,
  LabelWithCalloutProps
} from '../LabelWithCallout';
import Spinner from '../Spinner';
import ErrorMessage from '../ErrorMessage';

export enum FileFormatTypes {
  doc = '.doc',
  docx = '.docx',
  jpeg = '.jpeg',
  jpg = '.jpg',
  pdf = '.pdf',
  png = '.png',
  tif = '.tif',
  txt = '.txt',
  xml = '.xml'
}

export interface AttachmentMetadata extends File {
  id: string;
}

export interface FileUploaderProps {
  /** Akksepterte filformater */
  acceptedFileFormats?: Array<FileFormatTypes>;
  /** Tekst for opplastingskomponenten */
  addFileString?: string | JSX.Element;
  /** aria-label */
  ariaLabel?: string;
  /** string for Apikall */
  axiosPath?: string;
  /** CSS class */
  className?: string;
  /** Lukk callout på blur */
  labelWithCalloutAutoDismiss?: boolean;
  /** trigger funksjon til å slette alle filer */
  deleteAllFiles?: boolean;
  /** Aria-label for "fjern fil"-knapp */
  deleteButtonAriaLabel?: string;
  /** Funksjon for å slette opplastet fil */
  deleteFile?: (file: File) => void;
  /** Opplastede filer */
  files?: Array<any>;
  /** Hjelpetekst */
  help?: string | JSX.Element;
  /** Id */
  id?: string;
  /** Tilleggsinformasjon */
  info?: string;
  /** Descriptive label for SearchField */
  label?: string;
  /** aria-label for knapp i label */
  labelButtonAriaLabel?: string;
  /** Overstyr label, se LabelWithCallout komponent */
  labelCallout?: LabelWithCalloutProps;
  /** Spinner når fil laster */
  loading?: boolean;
  /** Mulighet for å laste opp flere filer */
  multipleFiles?: boolean;
  /** Brukerspesifisert event for callout **/
  onCalloutToggle?: (
    oldCalloutState: calloutState,
    newCalloutState: calloutState
  ) => void;
  /** Parametere for axioskall */
  queryParams?: any;
  /** Funksjon for filopplasting */
  uploadFile?: (file: File) => void;
  /**forsinkelse før opplasting i millisekunder*/
  forsinkelse?: number;
  /**erstatter tegn som er ugyldige */
  normalizeFileName?: boolean;
  /**Definer ugyldige tegn som skal erstattes med "_". Skal erstatte alle non-ord karakter dersom invalidCharacterRegexp ikker er oppgitt*/
  invalidCharacterRegexp?: RegExp;
  /**størrelsesgrense til en enkelt fil i bit*/
  fileSizeLimit?: number;
  /*feilmelding for oversteget av filstørrelsesgrense**/
  exceedFileSizeLimitErrorMessage?: string;
  /**overstyr intern feilmelding*/
  errorMessage?: string;
  /**Funkjson henrettet etter opplasting*/
  afterUpload?: () => void;
}

export const isCorrectFileFormat = (
  file: File,
  acceptedFilformats: Array<FileFormatTypes> | undefined
) => {
  if (!acceptedFilformats) {
    return true;
  }
  const fileExtention = file.name.match(/\.[0-9a-z]+$/i);
  if (fileExtention && fileExtention[0]) {
    if (acceptedFilformats.indexOf(fileExtention[0] as FileFormatTypes) > -1) {
      return true;
    }
  }
  return false;
};

const nonWordCharacterRegexp = /\W/g;
const fileNameRegex = /\.(?=[^.]+$)/;

const normalize = (file: File, invalidCharacterRegexp?: RegExp) => {
  const nameList = file.name.split(fileNameRegex);
  const fileName = nameList[0];
  const normalizedName = fileName.replace(
    invalidCharacterRegexp || nonWordCharacterRegexp,
    '_'
  );
  return normalizedName.concat('.', nameList[1]);
};

const FileUploader: React.FC<FileUploaderProps> = props => {
  const {
    acceptedFileFormats,
    addFileString,
    ariaLabel,
    axiosPath,
    className,
    labelWithCalloutAutoDismiss,
    deleteAllFiles,
    deleteButtonAriaLabel,
    deleteFile,
    files,
    help,
    id,
    info,
    label,
    labelButtonAriaLabel,
    labelCallout,
    loading,
    multipleFiles,
    onCalloutToggle,
    queryParams,
    uploadFile,
    normalizeFileName,
    invalidCharacterRegexp,
    fileSizeLimit,
    exceedFileSizeLimitErrorMessage,
    errorMessage,
    afterUpload
  } = props;
  const styles = getClassNames(props);
  const [internalFiles, setInternalFiles] = React.useState<Array<any>>(
    files ? files : []
  );
  const [internalErrorMessages, setInternalErrorMessages] = React.useState<
    string[]
  >([]);
  const [internalLoading, setInternalLoading] = React.useState<boolean>(false);
  const inputRef = React.useRef<HTMLInputElement | null>(null);

  React.useEffect(() => {
    if (files) {
      setInternalErrorMessages([]);
      setInternalFiles(files);
    }
  }, [files]);

  React.useEffect(() => {
    if (errorMessage) {
      setInternalErrorMessages([errorMessage]);
    }
  }, [errorMessage]);

  const pushToInternalMessages = (msg: string) =>
    setInternalErrorMessages(prevState => [...prevState, msg]);

  const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    event.stopPropagation();
    if (event.target.files && event.target.files.length > 0) {
      handleNewFiles(Array.from(event.target.files));
    }
  };

  const uploadFilePromise = (
    url: string,
    file: File,
    params?: any
  ): Promise<any> => {
    const formData = new FormData();
    formData.append(
      'upload',
      file,
      normalizeFileName ? normalize(file, invalidCharacterRegexp) : undefined
    );
    return axios.post<FormData, AxiosResponse<AttachmentMetadata>>(
      url,
      formData,
      { params }
    );
  };

  const isValidFile = (file: File, sizeLimit?: number) =>
    isCorrectFileFormat(file, acceptedFileFormats) &&
    (sizeLimit ? file.size <= sizeLimit : true);

  const handleNewFiles = (fileList: File[]) => {
    setInternalErrorMessages([]);

    const exceedSizeLimitFiles = fileSizeLimit
      ? fileList.filter(file => file.size > fileSizeLimit)
      : [];
    if (fileSizeLimit && exceedSizeLimitFiles.length) {
      pushToInternalMessages(
        exceedFileSizeLimitErrorMessage ||
          createDefaultOversizedFileErrorMessage(fileSizeLimit)
      );
    }

    const invalidFileFormatFiles = fileList.filter(
      file => !isCorrectFileFormat(file, acceptedFileFormats)
    );
    if (invalidFileFormatFiles.length) {
      pushToInternalMessages('Dette filformatet er ikke godkjent');
    }

    const validFiles = fileList.filter(file =>
      isValidFile(file, fileSizeLimit)
    );

    if (uploadFile) {
      validFiles.forEach(file => {
        uploadFile(file);
      });
    }

    if (!axiosPath) {
      return;
    }

    setInternalLoading(true);
    const allPromises = validFiles.map((file: File) =>
      uploadFilePromise(axiosPath, file, queryParams)
    );

    setTimeout(() => {
      axios
        .all(allPromises)
        .then(responses => {
          setInternalFiles([
            ...internalFiles,
            ...responses.map(res => res.data)
          ]);
        })
        .catch(errors => {
          //TODO: Det trenger design om flere feilmeldinger
          pushToInternalMessages('Kunne ikke laste opp fil');
        })
        .finally(() => {
          if (afterUpload) {
            afterUpload();
          }
          setInternalLoading(false);
        });
    }, props.forsinkelse || 0);
  };

  const createDefaultOversizedFileErrorMessage = (
    filstoerrelsegrense: number
  ) =>
    `Vi kan ikke motta denne filen fordi den er for stor. Filer kan ikke overstige ${bitToMegabyte(
      filstoerrelsegrense
    )} Mb. Du kan forsøke å dele opp i flere mindre filer, eller bruke et format som tar mindre plass.`;

  const bitToMegabyte = (size: number) => (size / (1024 * 1024)).toFixed(1);

  const handleDragOverAndDragEnter = (
    event: React.DragEvent<HTMLDivElement>
  ) => {
    event.preventDefault();
    event.stopPropagation();
  };

  const handleDragLeave = (event: React.DragEvent<HTMLDivElement>) => {
    event.preventDefault();
    event.stopPropagation();
  };

  const handleDrop = (event: React.DragEvent<HTMLDivElement>) => {
    event.preventDefault();
    event.stopPropagation();
    if (
      event.dataTransfer &&
      event.dataTransfer.files &&
      event.dataTransfer.files.length > 0
    ) {
      handleNewFiles(Array.from(event.dataTransfer.files));
    }
  };

  const deleteFromList = (fileToBeDeleted: AttachmentMetadata) => {
    if (deleteFile) {
      deleteFile(fileToBeDeleted);
    }
    if (axiosPath) {
      axios
        .delete(`${axiosPath}/${fileToBeDeleted.id}`, {
          params: queryParams,
          data: {} // kreves av BigIP
        })
        .then(() => {
          const newList = internalFiles.filter(
            f => f.id !== fileToBeDeleted.id
          );
          setInternalFiles(newList);
        });
    }
  };
  if (deleteAllFiles && files) {
    files.forEach(file => {
      deleteFromList(file);
    });
  }

  return (
    <div className={classnames(styles.main, className)}>
      <LabelWithCallout
        id={id}
        label={label}
        buttonAriaLabel={labelButtonAriaLabel}
        help={help}
        onCalloutToggle={onCalloutToggle}
        autoDismiss={labelWithCalloutAutoDismiss}
        {...labelCallout}
      />
      <label
        htmlFor="fileupload"
        aria-label={ariaLabel ? ariaLabel : 'Filopplasting'}
        id="buttonLabel"
      >
        <div
          className={styles.uploadArea}
          role="button"
          aria-describedby="acceptedFileFormats"
          tabIndex={0}
          onDragEnter={handleDragOverAndDragEnter}
          onDragLeave={handleDragLeave}
          onDragOver={handleDragOverAndDragEnter}
          onDrop={handleDrop}
          onClick={event => {
            event.preventDefault();
            if (inputRef.current) {
              inputRef.current.click();
            }
          }}
          onKeyPress={ev => {
            if (ev.keyCode === 0 && inputRef.current) {
              inputRef.current.click();
            }
          }}
        >
          {loading || internalLoading ? (
            <Spinner />
          ) : (
            <>
              <Icon iconName={'AttachFile'} className={styles.uploadAreaIcon} />
              <u>{addFileString ? addFileString : 'Legg til fil(er)'}</u>
            </>
          )}
        </div>
      </label>
      <input
        className={styles.fileUploadInput}
        type="file"
        id="fileupload"
        ref={inputRef}
        multiple={multipleFiles}
        onChange={handleFileChange}
        tabIndex={-1}
      />

      {acceptedFileFormats && (
        <span className={styles.informationWrapper} id="acceptedFileFormats">
          Aksepterte filformater:{' '}
          <span className={styles.acceptedFileFormats}>
            {acceptedFileFormats.map(
              (fileFormat: FileFormatTypes, index: number) => {
                if (index === acceptedFileFormats.length - 1) {
                  return fileFormat;
                } else {
                  return fileFormat.concat(', ');
                }
              }
            )}
          </span>
        </span>
      )}
      {info && (
        <div
          className={styles.informationWrapper}
          id="information"
          aria-label={'informasjon'}
        >
          {info}
        </div>
      )}
      {internalErrorMessages &&
        internalErrorMessages.map(msg => (
          <div key={msg}>
            <ErrorMessage>{msg}</ErrorMessage>
          </div>
        ))}
      {internalFiles.length > 0 && (
        <div role="alert">
          <ul className={styles.fileList}>
            {internalFiles.map((file, index: number) => (
              <li key={file.name.concat(index.toString())}>
                <div className={styles.fileName}>{file.name}</div>
                {file.error ? (
                  <Icon iconName={'Error'} className={styles.errorColor} />
                ) : (
                  <button
                    className={styles.fileListCancelBtn}
                    onClick={() => deleteFromList(file)}
                    aria-label={
                      deleteButtonAriaLabel
                        ? deleteButtonAriaLabel
                        : 'Fjern fil'
                    }
                  >
                    <Icon iconName={'Cancel'} />
                  </button>
                )}
              </li>
            ))}
          </ul>
        </div>
      )}
    </div>
  );
};

export default FileUploader;
