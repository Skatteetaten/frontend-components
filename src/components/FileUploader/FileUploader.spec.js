import React from 'react';
import { matches } from './../utils/test-utils';
import { FileUploader } from './FileUploader';
import { FileFormatTypes, Language } from './FileUploader.types';
import { mount, shallow } from 'enzyme';

function oppsettShallow(props) {
  return shallow(<FileUploader {...props} />);
}

function oppsettFullDOM(props) {
  return mount(<FileUploader {...props} />);
}

describe('FileUploader komponent', () => {
  it('matcher snapshot', () => {
    matches(<FileUploader />);
  });
  it('skal rendre med riktig props', () => {
    const wrapper = oppsettShallow({
      acceptedFileFormats: [FileFormatTypes.jpg, FileFormatTypes.png],
      ariaLabel: 'Filopplaster',
      uploadFile: jest.fn(),
    });
    expect(wrapper.find('span').first().text()).toEqual(
      'Aksepterte filformater: .jpg, .png'
    );
  });
  it('skal rendre med info', () => {
    const wrapper = oppsettShallow({
      info:
        'Husk å sjekke for sensitive personopplysninger, og evt fjerne disse før du laster opp vedleggene.',
      ariaLabel: 'Filopplaster',
      uploadFile: jest.fn(),
    });
    expect(wrapper.find('div').last().text()).toEqual(
      'Husk å sjekke for sensitive personopplysninger, og evt fjerne disse før du laster opp vedleggene.'
    );
  });
  it('skal vise engelske tekster dersom engelsk språk er valgt', () => {
    const wrapper = oppsettShallow({
      acceptedFileFormats: [FileFormatTypes.jpg, FileFormatTypes.png],
      language: Language.en,
      uploadFile: jest.fn(),
    });
    expect(wrapper.find('u').first().text()).toEqual('Add file(s)');
    expect(wrapper.find('span').first().text()).toEqual(
      'Accepted file formats: .jpg, .png'
    );
  });
  it('skal kjøre uploadFile-funksjon når bruker laster opp fil', () => {
    const mockFunc = jest.fn();
    const wrapper = oppsettFullDOM({
      acceptedFileFormats: [FileFormatTypes.doc, FileFormatTypes.docx],
      ariaLabel: 'Filopplaster',
      axiosPath: 'http://localhost',
      language: Language.nb,
      uploadFile: () => mockFunc(),
    });
    wrapper.find('input').simulate('change', {
      target: {
        files: [{ name: 'tekst.doc' }],
      },
    });
    expect(mockFunc.mock.calls.length).toBe(1);
  });
  it('skal gi feilmelding dersom bruker laster opp feil filtype', () => {
    const mockFunc = jest.fn();
    const wrapper = oppsettFullDOM({
      acceptedFileFormats: [FileFormatTypes.doc, FileFormatTypes.docx],
      ariaLabel: 'Filopplaster',
      axiosPath: 'http://localhost',
      uploadFile: () => mockFunc(),
    });
    expect(wrapper.exists('ErrorMessage')).toEqual(false);
    wrapper.find('input').simulate('change', {
      target: {
        files: [{ name: 'tekst.jpeg' }],
      },
    });
    expect(wrapper.exists('ErrorMessage')).toEqual(true);
    expect(wrapper.find('ErrorMessage').text()).toEqual(
      'Dette filformatet er ikke godkjent'
    );
  });
  it('skal kjøre deleteFile dersom bruker trykker kryss i liste', () => {
    const mockFunc = jest.fn();
    const mockFuncDelete = jest.fn((file) => file);
    const wrapper = oppsettFullDOM({
      ariaLabel: 'Filopplaster',
      uploadFile: () => mockFunc(),
      files: [{ name: 'FilNavn.png', id: '123456789' }],
      deleteFile: (file, err) => mockFuncDelete(file),
    });
    expect(wrapper.find('li').length).toEqual(1);
    expect(wrapper.find('li').first().text()).toEqual('FilNavn.png');
    wrapper.find('li').first().find('button').simulate('click');
    expect(mockFuncDelete).toHaveBeenCalled();
    expect(mockFuncDelete).toHaveBeenCalledWith({
      id: '123456789',
      name: 'FilNavn.png',
    });
  });
  it('støtter csv format', () => {
    const mockFunc = jest.fn();
    const wrapper = oppsettFullDOM({
      acceptedFileFormats: [FileFormatTypes.csv],
      ariaLabel: 'Filopplaster',
      axiosPath: 'http://localhost',
      uploadFile: () => mockFunc(),
    });
    expect(wrapper.find('span').first().text()).toEqual(
      'Aksepterte filformater: .csv'
    );
    wrapper.find('input').simulate('change', {
      target: {
        files: [{ name: 'fil.csv' }],
      },
    });
    expect(wrapper.exists('ErrorMessage')).toEqual(false);
    expect(mockFunc.mock.calls.length).toBe(1);
  });
});
