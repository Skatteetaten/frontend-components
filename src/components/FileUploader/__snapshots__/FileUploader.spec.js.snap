// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploader komponent matcher snapshot 1`] = `
<div
  className=


>
  <label
    id="buttonLabel"
  >
    <div
      aria-describedby="acceptedFileFormats"
      className=

          {
            background-color: #cde1f9;
            border: 2px dashed #1362ae;
            cursor: pointer;
            margin-bottom: 8px;
            margin-top: 8px;
            padding-bottom: 16px;
            padding-left: 16px;
            padding-right: 16px;
            padding-top: 16px;
            text-align: center;
          }
      onClick={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <i
        aria-hidden={true}
        className=

            {
              -moz-osx-font-smoothing: grayscale;
              -webkit-font-smoothing: antialiased;
              color: #1362ae;
              display: inline-block;
              font-family: "Material Icons";
              font-size: 1.25rem;
              font-style: normal;
              font-weight: normal;
              speak: none;
            }
        data-icon-name="AttachFile"
      >
        Ôè¢
      </i>
      <u>
        Legg til fil(er)
      </u>
    </div>
  </label>
  <input
    aria-hidden={true}
    className=

        {
          opacity: 0;
          overflow: hidden;
          position: absolute;
          z-index: -1;
        }
    id="fileupload-input"
    onChange={[Function]}
    tabIndex={-1}
    type="file"
  />
</div>
`;
